cirq/study/flatten_expressions.py

```diff
@@ -0,0 +1,368 @@
+    expressions from circuits with single symbols and transforming parameter
+    sweeps to match.
+    """
+
+    def __new__(cls, *args, **kwargs):
+        """Disables the behavior of `ParamResolver.__new__`."""
+        return super().__new__(cls)
+
+    def __init__(
+            self,
+            param_dict: Optional[resolver.ParamResolverOrSimilarType] = None,
+            *,  # Force keyword args
+            get_param_name: Callable[[
+                sympy.Basic,
+            ], str] = None):
+        """Initializes a new _ParamFlattener.
+
+        Args:
+            param_dict: A default initial mapping from some parameter names,
+                symbols, or expressions to other symbols or values.  Only sympy
+                expressions and symbols not specified in `param_dict` will be
+                flattened.
+            get_param_name: A callback function that returns a new parameter
+                name for a given sympy expression or symbol.  If this function
+                returns the same value for two different expressions, `'_#'` is
+                appended to the name to avoid name collision wher `#` is the
+                number of previous collisions.  By default, returns the
+                expression string surrounded by angle brackets e.g. `'<x+1>'`.
+        """
+        if hasattr(self, '_taken_symbols'):
+            # Has already been initialized
+            return
+        if isinstance(param_dict, resolver.ParamResolver):
+            params = param_dict.param_dict
+        else:
+            params = param_dict if param_dict else {}
+        symbol_params = {
+            _ensure_not_str(param): _ensure_not_str(val)
+            for param, val in params.items()
+        }
+        super().__init__(symbol_params)
+        if get_param_name is None:
+            get_param_name = self.default_get_param_name
+        self.get_param_name = get_param_name
+        self._taken_symbols = set(self.param_dict.values())
+
+    @staticmethod
+    def default_get_param_name(val: sympy.Basic) -> str:
+        if isinstance(val, sympy.Symbol):
+            return val.name
+        return '<{!s}>'.format(val)
+
+    def _next_symbol(self, val: sympy.Basic) -> sympy.Symbol:
+        name = self.get_param_name(val)
+        symbol = sympy.Symbol(name)
+        # Ensure the symbol hasn't already been used
+        collision = 0
+        while symbol in self._taken_symbols:
+            collision += 1
+            symbol = sympy.Symbol('{}_{}'.format(name, collision))
+        return symbol
+
+    def value_of(self, value: Union[sympy.Basic, float, str]
+                ) -> Union[sympy.Basic, float]:
+        """Resolves a symbol or expression to a new symbol unique to that value.
+
+        If value is a float, returns it.
+        If value is a str, treat it as a symbol with that name and continue.
+        If this `_ParamFlattener` was initialized with a `param_dict` and
```

Comment:
I would get rid of the "was initialized with a param_dict", since this returns param_dict[value] if it has been added by a previous call.

Reply:
Done.